### PR TITLE
[NODE][REFACTOR] Rename IRFunctor->NodeFunctor, use func pointer

### PR DESCRIPTION
**Pull request by tqchen**
Previously we used std::function for the functor dispatching.
It introduces additional overhead and problems during dll destruction(of std::function).

This PR changes the std::function to function pointers.
This change a bit restrictions around the set_dispatch that we can get around,
but will improve the general efficiency by reducing one level of indirection in the std::function.
We also no longer need special marcos to register functions to the Functor.

### CODE REVIEW

**Comment by tqchen** [comment]
cc @zhiics @jroesch @icemelon9

File: include/tvm/node/functor.h [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+  }
+  /*!
+   * \brief invoke the functor , dispatch on type of n
+   * \param n The Node argument
+   * \param args The additional arguments
+   * \return The result.
+   */
+  R operator()(const ObjectRef& n, Args... args) const {
+    uint32_t type_index = n->type_index();
+    CHECK(type_index < func_.size() &&
```
Comment (zhiics) [inline include/tvm/node/functor.h]: CHECK(can_dispath(n))?

File: include/tvm/node/functor.h [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+   * \brief Whether the functor can dispatch the corresponding Node
+   * \param n The node to be dispatched
+   * \return Whether dispatching function is registered for n's type.
+   */
+  bool can_dispatch(const ObjectRef& n) const {
+    uint32_t type_index = n->type_index();
+    return type_index < func_.size() && func_[type_index] != nullptr;
+  }
+  /*!
+   * \brief invoke the functor , dispatch on type of n
```
Comment (zhiics) [inline include/tvm/node/functor.h]:
```suggestion
   * \brief invoke the functor, dispatch on type of n
```

File: include/tvm/node/functor.h [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+#define TVM_REG_FUNC_VAR_DEF(ClsName)                                 \
+  static TVM_ATTRIBUTE_UNUSED auto & __make_functor ## _ ## ClsName
+
+/*!
+ * \brief Useful macro to set NodeFunctor dispatch in a global static field.
+ *
+ * \code
+ *  // Use NodeFunctor to implement IRPrinter similar to Visitor Pattern.
+ *  // vtable allows easy patch in of new Node types, without changing
```
Comment (zhiics) [inline include/tvm/node/functor.h]:
```suggestion
 *  // vtable allows easy patch of new Node types, without changing
```

**Comment by tqchen** [comment]
Thanks @zhiics  i have updated per your comment

File: include/tvm/node/functor.h [hunk 4]
```diff
@@ -84,15 +84,13 @@ class NodeFunctor<R(const ObjectRef& n, Args...)> {
...
    * \return The result.
    */
   R operator()(const ObjectRef& n, Args... args) const {
-    uint32_t type_index = n->type_index();
-    CHECK(type_index < func_.size() &&
-          func_[type_index] != nullptr)
+    CHECK(can_dispatch(n))
         << "NodeFunctor calls un-registered function on type "
         << n->GetTypeKey();
     return (*func_[type_index])(n, std::forward<Args>(args)...);
```
Comment (zhiics) [inline include/tvm/node/functor.h]: s/type_index/n->type_index()

This is probably the error.

**Comment by tqchen** [comment]
Thanks @zhiics

**Pull request merged**